WooCommerce/Classes/Authentication/Store Creation/StoreCreationCoordinator.swift

```diff
@@ -0,0 +1,135 @@
+        let webViewController = AuthenticatedWebViewController(viewModel: viewModel)
+        webViewController.addCloseNavigationBarButton(target: self, action: #selector(handleStoreCreationCloseAction))
+        let webNavigationController = WooNavigationController(rootViewController: webViewController)
+        // Disables interactive dismissal of the store creation modal.
+        webNavigationController.isModalInPresentation = true
+        navigationController.present(webNavigationController, animated: true)
+    }
+}
+
+private extension StoreCreationCoordinator {
+    func observeSiteURLsFromStoreCreation() {
+        possibleSiteURLsFromStoreCreationSubscription = $possibleSiteURLsFromStoreCreation
+            .filter { $0.isEmpty == false }
+            .removeDuplicates()
+            // There are usually three URLs in the webview that return a site URL - two with `*.wordpress.com` and the other the final URL.
+            .debounce(for: .seconds(5), scheduler: DispatchQueue.main)
+            .asyncMap { [weak self] possibleSiteURLs -> Site? in
+                // Waits for 5 seconds before syncing sites every time.
+                try await Task.sleep(nanoseconds: 5_000_000_000)
+                return try await self?.syncSites(forSiteThatMatchesPossibleURLs: possibleSiteURLs)
+            }
+            // Retries 10 times with 5 seconds pause in between to wait for the newly created site to be available as a Jetpack site
+            // in the WPCOM `/me/sites` response.
+            .retry(10)
+            .replaceError(with: nil)
+            .receive(on: DispatchQueue.main)
+            .sink { [weak self] site in
+                guard let self, let site else { return }
+                self.continueWithSelectedSite(site: site)
+            }
+    }
+
+    @objc func handleStoreCreationCloseAction() {
+        // TODO-7879: show a confirmation alert before closing the store creation view
+        // TODO-7879: analytics
+        navigationController.dismiss(animated: true)
+    }
+
+    func handleStoreCreationResult(_ result: Result<String, Error>) {
+        switch result {
+        case .success(let siteURL):
+            // TODO-7879: analytics
+
+            // There could be multiple site URLs from the completion URL in the webview, and only one
+            // of them matches the final site URL from WPCOM `/me/sites` endpoint.
+            possibleSiteURLsFromStoreCreation.insert(siteURL)
+        case .failure(let error):
+            // TODO-7879: analytics
+            DDLogError("Store creation error: \(error)")
+        }
+    }
+
+    @MainActor
+    func syncSites(forSiteThatMatchesPossibleURLs possibleURLs: Set<String>) async throws -> Site {
+        return try await withCheckedThrowingContinuation { [weak self] continuation in
+            storePickerViewModel.refreshSites(currentlySelectedSiteID: nil) { [weak self] in
+                guard let self else { return }
```

Comment:
`continuation` should be resumed here to avoid leak:
```suggestion
                guard let self else { return continuation.resume(throwing: <some unknown error> }
```

Reply:
updated in https://github.com/woocommerce/woocommerce-ios/pull/7927/commits/730370922a0fa28289685db85af6e2c461828c26 - I hope we don't have to handle such weak self all the time 😓